File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/semaphores/LockManager.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */
using System;
using com.db4o;
using com.db4o.ext;

namespace com.db4odoc.f1.semaphores
{
	/**
	 * This class demonstrates a very rudimentary implementation
	 * of virtual "locks" on objects with db4o. All code that is
	 * intended to obey these locks will have to call lock() and
	 * unlock().  
	 */
	public class LockManager 
	{
    
		readonly private string SEMAPHORE_NAME = "locked: ";
		readonly private int WAIT_FOR_AVAILABILITY = 300; // 300 milliseconds
    
		readonly private ExtObjectContainer _objectContainer;
    
		public LockManager(ObjectContainer objectContainer)
		{
			_objectContainer = objectContainer.Ext();
		}
    
		public bool Lock(object obj)
		{
			long id = _objectContainer.GetID(obj);
			return _objectContainer.SetSemaphore(SEMAPHORE_NAME + id, WAIT_FOR_AVAILABILITY);
		}
    
		public void Unlock(Object obj)
		{
			long id = _objectContainer.GetID(obj);
			_objectContainer.ReleaseSemaphore(SEMAPHORE_NAME + id);
		}
	}

}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/semaphores/LockManager.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 
Imports System
Imports com.db4o
Imports com.db4o.ext

Namespace com.db4odoc.f1.semaphores
    ' 	*
    ' 	 * This class demonstrates a very rudimentary implementation
    ' 	 * of virtual "locks" on objects with db4o. All code that is
    ' 	 * intended to obey these locks will have to call lock() and
    ' 	 * unlock().  
    ' 	 */
    Public Class LockManager

        Private ReadOnly SEMAPHORE_NAME As String = "locked: "
        Private ReadOnly WAIT_FOR_AVAILABILITY As Integer = 300  ' 300 milliseconds

        Private ReadOnly _objectContainer As ExtObjectContainer

        Public Sub New(ByVal objectContainer As ObjectContainer)
            _objectContainer = objectContainer.Ext()
        End Sub

        Public Function Lock(ByVal obj As Object) As Boolean
            Dim id As Long = _objectContainer.GetID(obj)
            Return _objectContainer.SetSemaphore(SEMAPHORE_NAME + id, WAIT_FOR_AVAILABILITY)
        End Function

        Public Sub Unlock(ByVal obj As Object)
            Dim id As Long = _objectContainer.GetID(obj)
            _objectContainer.ReleaseSemaphore(SEMAPHORE_NAME + id)
        End Sub
    End Class

End Namespace
